pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/impl/PersistentTopicsBase.java

```diff
@@ -2615,8 +2600,147 @@ protected PersistentOfflineTopicStats internalGetBacklog(boolean authoritative)
+            validateGlobalNamespaceOwnership(namespaceName);
+        }
+
+        PersistentTopic topic = (PersistentTopic) getTopicReference(topicName);
+        PositionImpl pos = new PositionImpl(messageId.getLedgerId(), messageId.getEntryId());
+
+        try {
+            ManagedLedgerImpl managedLedger = (ManagedLedgerImpl) topic.getManagedLedger();
+            synchronized (this) {
+                if(messageId.getLedgerId() == -1) {
+                    asyncResponse.resume(managedLedger.getTotalSize());
+                } else {
+                    NavigableMap<Long, MLDataFormats.ManagedLedgerInfo.LedgerInfo> ledgers = managedLedger.getLedgersInfo();
+                    MLDataFormats.ManagedLedgerInfo.LedgerInfo ledgerInfo = ledgers.get(pos.getLedgerId());
+                    if (ledgerInfo == null) {
+                        long sizeBeforePosLedger = ledgers.values().stream().filter(li -> li.getLedgerId() < pos.getLedgerId())
+                                .mapToLong(MLDataFormats.ManagedLedgerInfo.LedgerInfo::getSize).sum();
+                        long currentLedgerId = ledgers.lastKey();
+                        if (currentLedgerId < pos.getLedgerId()) {
+                            sizeBeforePosLedger += managedLedger.getCurrentLedgerSize();
+                        }
+                        asyncResponse.resume(managedLedger.getTotalSize() - sizeBeforePosLedger);
+                    } else {
+                        asyncResponse.resume(managedLedger.getEstimatedBacklogSize(pos));
+                    }
+                }
+
+            }
+        } catch (WebApplicationException wae) {
+            if (log.isDebugEnabled()) {
+                log.debug("[{}] Failed to get backlog size for non-partitioned topic {}, redirecting to other brokers.",
+                        clientAppId(), topicName, wae);
+            }
+            resumeAsyncResponseExceptionally(asyncResponse, wae);
+        } catch (Exception e) {
+            log.error("[{}] Failed to get backlog size for non-partitioned topic {}", clientAppId(), topicName, e);
+            resumeAsyncResponseExceptionally(asyncResponse, e);
+        }
+    }
+
+    private boolean isLedgerExists(MessageIdImpl messageId){
+        try {
+            String internalInfo = pulsar().getAdminClient().topics().getInternalInfo(topicName.toString());
+            boolean ledgerExists = false;
+            if(messageId.getLedgerId() == -1) {
+                ledgerExists = true;
+            } else {
+                if (topicName.isPartitioned()) {
+                    ManagedLedgerInfo managedLedgerInfo = jsonMapper().readValue(internalInfo, ManagedLedgerInfo.class);
+                    ledgerExists = managedLedgerInfo.ledgers.stream().anyMatch(ledgerInfo -> messageId.getLedgerId() == ledgerInfo.ledgerId);
+                } else {
+                    JsonNode partitionedInternalInfos = jsonMapper().readTree(internalInfo).get("partitions");
```

Comment:
seems we're processing partitioned topic here, should be` !topicName.isPartitioned()` in if clause?

Reply:
No, ```topicName.isPartitioned()``` means the topic is with partition, like 'topicname-partition-0'